internal/controller/etcdcluster_controller.go

```diff
@@ -36,24 +44,349 @@ type EtcdClusterReconciler struct {
+				"ETCD_INITIAL_CLUSTER_STATE": "new",
+			},
+		}
+		if err := ctrl.SetControllerReference(cluster, configMap, r.Scheme); err != nil {
+			return fmt.Errorf("cannot set controller reference: %w", err)
+		}
+		if err := r.Create(ctx, configMap); err != nil {
+			return fmt.Errorf("cannot create cluster state configmap: %w", err)
+		}
+		return nil
+	}
+	return fmt.Errorf("cannot get cluster state configmap: %w", err)
+}
+
+func (r *EtcdClusterReconciler) ensureClusterStatefulSet(
+	ctx context.Context, cluster *etcdaenixiov1alpha1.EtcdCluster) error {
+	statefulSet := &appsv1.StatefulSet{}
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}, statefulSet)
+
+	// statefulset does not exist, create new one
+	notFound := false
+	if errors.IsNotFound(err) {
+		notFound = true
+		// prepare initial cluster members
+		initialCluster := ""
+		for i := uint(0); i < cluster.Spec.Replicas; i++ {
+			if i > 0 {
+				initialCluster += ","
+			}
+			initialCluster += fmt.Sprintf("%s-%d=https://%s-%d.%s.%s.svc.cluster.local:2380",
+				cluster.Name, i,
+				cluster.Name, i, cluster.Name, cluster.Namespace,
+			)
+		}
+
+		statefulSet = &appsv1.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: cluster.Namespace,
+				Name:      cluster.Name,
+			},
+			Spec: appsv1.StatefulSetSpec{
+				// initialize static fields that cannot be changed across updates.
+				Replicas:            new(int32),
+				ServiceName:         cluster.Name,
+				PodManagementPolicy: appsv1.ParallelPodManagement,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app.kubernetes.io/name":       "etcd",
+						"app.kubernetes.io/instance":   cluster.Name,
+						"app.kubernetes.io/managed-by": "etcd-operator",
+					},
+				},
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{
+							"app.kubernetes.io/name":       "etcd",
+							"app.kubernetes.io/instance":   cluster.Name,
+							"app.kubernetes.io/managed-by": "etcd-operator",
+						},
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "etcd",
+								Image: "quay.io/coreos/etcd:v3.5.12",
+								Command: []string{
+									"etcd",
+									"--name=$(POD_NAME)",
+									"--listen-peer-urls=https://0.0.0.0:2380",
+									// for first version disable TLS for client access
+									"--listen-client-urls=http://0.0.0.0:2379",
+									"--initial-advertise-peer-urls=https://$(POD_NAME)." + cluster.Name + ".$(POD_NAMESPACE).svc.cluster.local:2380",
+									"--data-dir=/var/run/etcd/default.etcd",
+									"--initial-cluster=" + initialCluster,
+									fmt.Sprintf("--initial-cluster-token=%s-%s", cluster.Name, cluster.Namespace),
+									"--auto-tls",
+									"--peer-auto-tls",
+									"--advertise-client-urls=http://$(POD_NAME)." + cluster.Name + ".$(POD_NAMESPACE).svc.cluster.local:2379",
```

Comment:
Please do not hardcode `cluster.local` suffixes as many clusters may have different ones.

It's possible to access services without them.